File: android/app/src/main/java/com/lotto7/generator/Lotto7Engine.kt
package com.lotto7.generator

import kotlin.math.sqrt
import kotlin.random.Random

object Lotto7Engine {
    private const val MIN_NUM = 1
    private const val MAX_NUM = 37
    private const val MAIN_COUNT = 7

    fun kouOf(n: Int): Int = when {
        n <= 7 -> 1
        n <= 14 -> 2
        n <= 21 -> 3
        n <= 28 -> 4
        else -> 5
    }

    fun kouDistribution(nums: List<Int>): List<Int> {
        val counts = IntArray(6)
        nums.forEach { counts[kouOf(it)]++ }
        return (1..5).map { counts[it] }
    }

    class PatternAnalyzer(private val draws: List<Draw>) {
        val freq = mutableMapOf<Int, Int>()
        val monthFreq = mutableMapOf<Int, MutableMap<Int, Int>>()
        val kouPatterns = mutableMapOf<List<Int>, Int>()
        val oddPatterns = mutableMapOf<Int, Int>()
        val sums = mutableListOf<Int>()
        var sumAvg: Double = 0.0
        var sumStd: Double = 0.0
        lateinit var topKou: Set<List<Int>>
        lateinit var topOdd: Set<Int>

        init {
            analyze()
        }

        private fun analyze() {
            draws.forEach { draw ->
                draw.nums.forEach { n ->
                    freq[n] = (freq[n] ?: 0) + 1
                    draw.month?.let { month ->
                        val monthMap = monthFreq.getOrPut(month) { mutableMapOf() }
                        monthMap[n] = (monthMap[n] ?: 0) + 1
                    }
                }
                val kou = kouDistribution(draw.nums)
                kouPatterns[kou] = (kouPatterns[kou] ?: 0) + 1
                val odd = draw.nums.count { it % 2 == 1 }
                oddPatterns[odd] = (oddPatterns[odd] ?: 0) + 1
                sums.add(draw.nums.sum())
            }

            sumAvg = sums.average()
            sumStd = sqrt(sums.map { (it - sumAvg) * (it - sumAvg) }.average())
            topKou = kouPatterns.entries
                .sortedByDescending { it.value }
                .take(30)
                .map { it.key }
                .toSet()
            topOdd = oddPatterns.entries
                .sortedByDescending { it.value }
                .take(4)
                .map { it.key }
                .toSet()
        }

        fun numberWeights(
            targetMonth: Int? = null,
            savedWinningNumbers: List<List<Int>> = emptyList()
        ): Map<Int, Double> {
            val weights = mutableMapOf<Int, Double>()
            for (n in MIN_NUM..MAX_NUM) {
                weights[n] = (freq[n] ?: 1).toDouble()
            }
            targetMonth?.let { month ->
                monthFreq[month]?.forEach { (n, count) ->
                    weights[n] = (weights[n] ?: 0.0) + count * 0.5
                }
            }
            if (savedWinningNumbers.isNotEmpty()) {
                val recentFreq = mutableMapOf<Int, Int>()
                savedWinningNumbers.take(20).forEach { nums ->
                    nums.forEach { n -> recentFreq[n] = (recentFreq[n] ?: 0) + 1 }
                }
                recentFreq.forEach { (n, count) ->
                    weights[n] = (weights[n] ?: 1.0) * (1.0 - count * 0.12).coerceAtLeast(0.3)
                }
                val allWinning = savedWinningNumbers.flatten().toSet()
                for (n in MIN_NUM..MAX_NUM) {
                    if (n !in allWinning) {
                        weights[n] = (weights[n] ?: 1.0) * 1.08
                    }
                }
            }
            return weights
        }

        fun buildSummary(targetMonth: Int?): AnalysisSummary {
            val hot = freq.entries.sortedByDescending { it.value }.take(5).map { it.key }
            val cold = freq.entries.sortedBy { it.value }.take(5).map { it.key }
            val topOdd = oddPatterns.entries.sortedByDescending { it.value }.take(3)
                .map { it.key to it.value }
            val latest = draws.last()

            return AnalysisSummary(
                totalDraws = draws.size,
                latestRound = latest.round,
                latestDate = latest.date,
                latestNums = latest.nums,
                hotNumbers = hot,
                coldNumbers = cold,
                sumAverage = sumAvg,
                sumMin = sums.minOrNull() ?: 0,
                sumMax = sums.maxOrNull() ?: 0,
                topOddPatterns = topOdd,
                targetMonth = targetMonth
            )
        }
    }

    class Generator(
        private val analyzer: PatternAnalyzer,
        private val random: Random = Random.Default,
        private val savedWinningNumbers: List<List<Int>> = emptyList()
    ) {
        private fun overlapsWinning(nums: List<Int>): Boolean {
            if (savedWinningNumbers.isEmpty()) return false
            if (nums in savedWinningNumbers) return true
            return savedWinningNumbers.any { winning ->
                nums.intersect(winning.toSet()).size >= 5
            }
        }

        private fun fitsPattern(nums: List<Int>): Boolean {
            if (overlapsWinning(nums)) return false
            val kou = kouDistribution(nums)
            val odd = nums.count { it % 2 == 1 }
            val total = nums.sum()
            if (kou !in analyzer.topKou) return false
            if (odd !in analyzer.topOdd) return false
            if (total !in (analyzer.sumAvg - 1.8 * analyzer.sumStd).toInt()
                    ..(analyzer.sumAvg + 1.8 * analyzer.sumStd).toInt()) return false
            return true
        }

        private fun weightedPick(weights: Map<Int, Double>, exclude: Set<Int>): Int {
            val candidates = (MIN_NUM..MAX_NUM).filter { it !in exclude }
            val totalWeight = candidates.sumOf { weights[it] ?: 1.0 }
            var roll = random.nextDouble() * totalWeight
            for (n in candidates) {
                roll -= weights[n] ?: 1.0
                if (roll <= 0) return n
            }
            return candidates.last()
        }

        fun generateOne(targetMonth: Int? = null): List<Int> {
            val weights = analyzer.numberWeights(targetMonth, savedWinningNumbers)
            repeat(500) {
                val nums = mutableSetOf<Int>()
                while (nums.size < MAIN_COUNT) {
                    nums.add(weightedPick(weights, nums))
                }
                val result = nums.sorted()
                if (fitsPattern(result)) return result
            }
            val nums = mutableSetOf<Int>()
            while (nums.size < MAIN_COUNT) {
                nums.add(weightedPick(weights, nums))
            }
            return nums.sorted()
        }

        fun generate(count: Int = 10, targetMonth: Int? = null): List<List<Int>> {
            val seen = mutableSetOf<List<Int>>()
            val results = mutableListOf<List<Int>>()
            var attempts = 0
            while (results.size < count && attempts < count * 200) {
                attempts++
                val nums = generateOne(targetMonth)
                if (seen.add(nums)) {
                    results.add(nums)
                }
            }
            return results
        }
    }

    fun resolveTargetMonth(draws: List<Draw>, overrideMonth: Int? = null): Int? {
        if (overrideMonth != null) return overrideMonth.coerceIn(1, 12)
        val lastMonth = draws.lastOrNull()?.month ?: return null
        return lastMonth % 12 + 1
    }
}

File: src/engine/Lotto7Engine.ts
import type { AnalysisSummary, Draw } from '../models/types';

const MIN_NUM = 1;
const MAX_NUM = 37;
const MAIN_COUNT = 7;

export function kouOf(n: number): number {
  if (n <= 7) return 1;
  if (n <= 14) return 2;
  if (n <= 21) return 3;
  if (n <= 28) return 4;
  return 5;
}

export function kouDistribution(nums: number[]): number[] {
  const counts = [0, 0, 0, 0, 0, 0];
  nums.forEach((n) => {
    counts[kouOf(n)]++;
  });
  return counts.slice(1);
}

export class PatternAnalyzer {
  freq: Record<number, number> = {};
  monthFreq: Record<number, Record<number, number>> = {};
  kouPatterns: Record<string, number> = {};
  oddPatterns: Record<number, number> = {};
  sums: number[] = [];
  sumAvg = 0;
  sumStd = 0;
  topKou = new Set<string>();
  topOdd = new Set<number>();

  constructor(private draws: Draw[]) {
    this.analyze();
  }

  private analyze() {
    this.draws.forEach((draw) => {
      draw.nums.forEach((n) => {
        this.freq[n] = (this.freq[n] ?? 0) + 1;
        if (draw.month != null) {
          if (!this.monthFreq[draw.month]) this.monthFreq[draw.month] = {};
          this.monthFreq[draw.month][n] = (this.monthFreq[draw.month][n] ?? 0) + 1;
        }
      });
      const kou = kouDistribution(draw.nums);
      const kouKey = kou.join(',');
      this.kouPatterns[kouKey] = (this.kouPatterns[kouKey] ?? 0) + 1;
      const odd = draw.nums.filter((n) => n % 2 === 1).length;
      this.oddPatterns[odd] = (this.oddPatterns[odd] ?? 0) + 1;
      this.sums.push(draw.nums.reduce((a, b) => a + b, 0));
    });

    this.sumAvg = this.sums.reduce((a, b) => a + b, 0) / this.sums.length;
    const variance =
      this.sums.reduce((acc, s) => acc + (s - this.sumAvg) ** 2, 0) / this.sums.length;
    this.sumStd = Math.sqrt(variance);

    this.topKou = new Set(
      Object.entries(this.kouPatterns)
        .sort((a, b) => b[1] - a[1])
        .slice(0, 30)
        .map(([k]) => k)
    );
    this.topOdd = new Set(
      Object.entries(this.oddPatterns)
        .sort((a, b) => b[1] - a[1])
        .slice(0, 4)
        .map(([k]) => Number(k))
    );
  }

  numberWeights(
    targetMonth?: number | null,
    savedWinningNumbers: number[][] = []
  ): Record<number, number> {
    const weights: Record<number, number> = {};
    for (let n = MIN_NUM; n <= MAX_NUM; n++) {
      weights[n] = this.freq[n] ?? 1;
    }
    if (targetMonth != null && this.monthFreq[targetMonth]) {
      Object.entries(this.monthFreq[targetMonth]).forEach(([n, count]) => {
        weights[Number(n)] = (weights[Number(n)] ?? 0) + count * 0.5;
      });
    }
    if (savedWinningNumbers.length > 0) {
      const recentFreq: Record<number, number> = {};
      savedWinningNumbers.slice(0, 20).forEach((nums) => {
        nums.forEach((n) => {
          recentFreq[n] = (recentFreq[n] ?? 0) + 1;
        });
      });
      Object.entries(recentFreq).forEach(([n, count]) => {
        weights[Number(n)] = (weights[Number(n)] ?? 1) * Math.max(1 - count * 0.12, 0.3);
      });
      const allWinning = new Set(savedWinningNumbers.flat());
      for (let n = MIN_NUM; n <= MAX_NUM; n++) {
        if (!allWinning.has(n)) {
          weights[n] = (weights[n] ?? 1) * 1.08;
        }
      }
    }
    return weights;
  }

  buildSummary(targetMonth: number | null): AnalysisSummary {
    const hot = Object.entries(this.freq)
      .sort((a, b) => b[1] - a[1])
      .slice(0, 5)
      .map(([n]) => Number(n));
    const cold = Object.entries(this.freq)
      .sort((a, b) => a[1] - b[1])
      .slice(0, 5)
      .map(([n]) => Number(n));
    const topOdd = Object.entries(this.oddPatterns)
      .sort((a, b) => b[1] - a[1])
      .slice(0, 3)
      .map(([k, v]) => [Number(k), v] as [number, number]);
    const latest = this.draws[this.draws.length - 1];

    return {
      totalDraws: this.draws.length,
      latestRound: latest.round,
      latestDate: latest.date,
      latestNums: latest.nums,
      hotNumbers: hot,
      coldNumbers: cold,
      sumAverage: this.sumAvg,
      sumMin: Math.min(...this.sums),
      sumMax: Math.max(...this.sums),
      topOddPatterns: topOdd,
      targetMonth,
    };
  }
}

export class Generator {
  constructor(
    private analyzer: PatternAnalyzer,
    private savedWinningNumbers: number[][] = []
  ) {}

  private overlapsWinning(nums: number[]): boolean {
    if (this.savedWinningNumbers.length === 0) return false;
    const key = nums.join(',');
    if (this.savedWinningNumbers.some((w) => w.join(',') === key)) return true;
    const set = new Set(nums);
    return this.savedWinningNumbers.some((winning) => {
      const overlap = winning.filter((n) => set.has(n)).length;
      return overlap >= 5;
    });
  }

  private fitsPattern(nums: number[]): boolean {
    if (this.overlapsWinning(nums)) return false;
    const kou = kouDistribution(nums);
    const kouKey = kou.join(',');
    const odd = nums.filter((n) => n % 2 === 1).length;
    const total = nums.reduce((a, b) => a + b, 0);
    if (!this.analyzer.topKou.has(kouKey)) return false;
    if (!this.analyzer.topOdd.has(odd)) return false;
    const min = Math.floor(this.analyzer.sumAvg - 1.8 * this.analyzer.sumStd);
    const max = Math.ceil(this.analyzer.sumAvg + 1.8 * this.analyzer.sumStd);
    if (total < min || total > max) return false;
    return true;
  }

  private weightedPick(weights: Record<number, number>, exclude: Set<number>): number {
    const candidates = Array.from({ length: MAX_NUM - MIN_NUM + 1 }, (_, i) => i + MIN_NUM).filter(
      (n) => !exclude.has(n)
    );
    const totalWeight = candidates.reduce((sum, n) => sum + (weights[n] ?? 1), 0);
    let roll = Math.random() * totalWeight;
    for (const n of candidates) {
      roll -= weights[n] ?? 1;
      if (roll <= 0) return n;
    }
    return candidates[candidates.length - 1];
  }

  generateOne(targetMonth?: number | null): number[] {
    const weights = this.analyzer.numberWeights(targetMonth, this.savedWinningNumbers);
    for (let i = 0; i < 500; i++) {
      const nums = new Set<number>();
      while (nums.size < MAIN_COUNT) {
        nums.add(this.weightedPick(weights, nums));
      }
      const result = Array.from(nums).sort((a, b) => a - b);
      if (this.fitsPattern(result)) return result;
    }
    const nums = new Set<number>();
    while (nums.size < MAIN_COUNT) {
      nums.add(this.weightedPick(weights, nums));
    }
    return Array.from(nums).sort((a, b) => a - b);
  }

  generate(count = 10, targetMonth?: number | null): number[][] {
    const seen = new Set<string>();
    const results: number[][] = [];
    let attempts = 0;
    while (results.length < count && attempts < count * 200) {
      attempts++;
      const nums = this.generateOne(targetMonth);
      const key = nums.join(',');
      if (!seen.has(key)) {
        seen.add(key);
        results.push(nums);
      }
    }
    return results;
  }
}

export function resolveTargetMonth(draws: Draw[], overrideMonth?: number | null): number | null {
  if (overrideMonth != null) return Math.min(12, Math.max(1, overrideMonth));
  const lastMonth = draws[draws.length - 1]?.month;
  if (lastMonth == null) return null;
  return (lastMonth % 12) + 1;
}
